### PR TITLE
HL-15

### PR DESCRIPTION
**Pull request by OptimusOpus**
[Link to ticket](https://dltx.atlassian.net/jira/software/projects/HL/boards/82?selectedIssue=HL-15)

### CODE REVIEW

File: contracts/VaultTimeLock.sol [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+    /**
+     * @dev Deposit/mint common workflow.
+     */
+    function _deposit(
+        address caller,
+        address receiver,
+        uint256 assets,
+        uint256 shares
+    ) internal virtual override {
+        lockedTime[receiver] = block.timestamp + lockDuration;
```
Comment (brettvniekerk) [inline contracts/VaultTimeLock.sol:35]: shouldn't this be a require statement? also shouldn't there be a require statement for `receiver` being the owner?

File: contracts/VaultTimeLock.sol [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+     * @dev Withdraw/redeem common workflow.
+     */
+    function _withdraw(
+        address caller,
+        address receiver,
+        address owner,
+        uint256 assets,
+        uint256 shares
+    ) internal virtual override {
+        require(lockedTime[owner] < block.timestamp, "_withdraw: Locked time not passed");
```
Comment (brettvniekerk) [inline contracts/VaultTimeLock.sol:49]: do we also want the receiver of withdrawn assets to be the original depositer as well?

File: contracts/VaultTimeLock.sol [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+        return super._withdraw(caller, receiver, owner, assets, shares);
+    }
+
+    // transfer/transferFrom common workflow.
+    // Override _transfer function to require lock time to be passed
+    // This is what prevents users bypassing withdrawals before lock period is over
+    function _transfer(
```
Comment (brettvniekerk) [inline contracts/VaultTimeLock.sol:56]: does this prevent the transfer of the share token from one account to the other?

Comment (OptimusOpus) [reply contracts/VaultTimeLock.sol:56]: Yep

File: contracts/VaultTimeLock.sol (hunk 1 above)
Comment (bitcoinbrisbane) [reply contracts/VaultTimeLock.sol:35]: You can deposit to someone elese account

**Review by bitcoinbrisbane — CHANGES_REQUESTED** [review]
Need to fix unit tests

**Comment by bitcoinbrisbane** [comment]
```bash
1) Market
       "before each" hook for "should have properties set on deploy":
     ERROR processing /home/lucascullen/GitHub/horse-link/contracts.horse.link/deploy/040_Vault.ts:
Error: invalid BigNumber value (argument="value", value=undefined, code=INVALID_ARGUMENT, version=bignumber/5.7.0)
    at Logger.makeError (/home/lucascullen/GitHub/horse-link/contracts.horse.link/node_modules/@ethersproject/logger/src.ts/index.ts:269:28)
    at Logger.throwError (/home/lucascullen/GitHub/horse-link/contracts.horse.link/node_modules/@ethersproject/logger/src.ts/index.ts:281:20)
    at Logger.throwArgumentError (/home/lucascullen/GitHub/horse-link/contracts.horse.link/node_modules/@ethersproject/logger/src.ts/index.ts:285:21)
    at Function.BigNumber.from (/home/lucascullen/GitHub/horse-link/contracts.horse.link/node_modules/@ethersproject/bignumber/src.ts/bignumber.ts:289:23)
    at NumberCoder.encode (/home/lucascullen/GitHub/horse-link/contracts.horse.link/node_modules/@ethersproject/abi/src.ts/coders/number.ts:25:27)
    at /home/lucascullen/GitHub/horse-link/contracts.horse.link/node_modules/@ethersproject/abi/src.ts/coders/array.ts:71:19
    at Array.forEach (<anonymous>)
    at pack (/home/lucascullen/GitHub/horse-link/contracts.horse.link/node_modules/@ethersproject/abi/src.ts/coders/array.ts:54:12)
    at TupleCoder.encode (/home/lucascullen/GitHub/horse-link/contracts.horse.link/node_modules/@ethersproject/abi/src.ts/coders/tuple.ts:54:20)
    at AbiCoder.encode (/home/lucascullen/GitHub/horse-link/contracts.horse.link/node_modules/@ethersproject/abi/src.ts/abi-coder.ts:111:15)
  Error: ERROR processing /home/lucascullen/GitHub/horse-link/contracts.horse.link/deploy/040_Vault.ts:
  Error: invalid BigNumber value (argument="value", value=undefined, code=INVALID_ARGUMENT, version=bignumber/5.7.0)
      at Logger.makeError (node_modules/@ethersproject/logger/src.ts/index.ts:269:28)
      at Logger.throwError (node_modules/@ethersproject/logger/src.ts/index.ts:281:20)
      at Logger.throwArgumentError (node_modules/@ethersproject/logger/src.ts/index.ts:285:21)
      at Function.BigNumber.from (node_modules/@ethersproject/bignumber/src.ts/bignumber.ts:289:23)
      at NumberCoder.encode (node_modules/@ethersproject/abi/src.ts/coders/number.ts:25:27)
      at /home/lucascullen/GitHub/horse-link/contracts.horse.link/node_modules/@ethersproject/abi/src.ts/coders/array.ts:71:19
      at Array.forEach (<anonymous>)
      at pack (node_modules/@ethersproject/abi/src.ts/coders/array.ts:54:12)
      at TupleCoder.encode (node_modules/@ethersproject/abi/src.ts/coders/tuple.ts:54:20)
      at AbiCoder.encode (node_modules/@ethersproject/abi/src.ts/abi-coder.ts:111:15)
      at DeploymentsManager.executeDeployScripts (node_modules/hardhat-deploy/src/DeploymentsManager.ts:1222:19)
      at processTicksAndRejections (node:internal/process/task_queues:96:5)
      at runNextTicks (node:internal/process/task_queues:65:3)
      at listOnTimeout (node:internal/timers:528:9)
      at processTimers (node:internal/timers:502:7)
      at async DeploymentsManager.runDeploy (node_modules/hardhat-deploy/src/DeploymentsManager.ts:1052:5)
      at async Object.fixture (node_modules/hardhat-deploy/src/DeploymentsManager.ts:315:9)
      at async Context.<anonymous> (tests/market.test.ts:48:19)

  2) Registry
       "before each" hook for "Should have no markets or vaults":
     ERROR processing /home/lucascullen/GitHub/horse-link/contracts.horse.link/deploy/040_Vault.ts:
Error: invalid BigNumber value (argument="value", value=undefined, code=INVALID_ARGUMENT, version=bignumber/5.7.0)
    at Logger.makeError (/home/lucascullen/GitHub/horse-link/contracts.horse.link/node_modules/@ethersproject/logger/src.ts/index.ts:269:28)
    at Logger.throwError (/home/lucascullen/GitHub/horse-link/contracts.horse.link/node_modules/@ethersproject/logger/src.ts/index.ts:281:20)
    at Logger.throwArgumentError (/home/lucascullen/GitHub/horse-link/contracts.horse.link/node_modules/@ethersproject/logger/src.ts/index.ts:285:21)
    at Function.BigNumber.from (/home/lucascullen/GitHub/horse-link/contracts.horse.link/node_modules/@ethersproject/bignumber/src.ts/bignumber.ts:289:23)
    at NumberCoder.encode (/home/lucascullen/GitHub/horse-link/contracts.horse.link/node_modules/@ethersproject/abi/src.ts/coders/number.ts:25:27)
    at /home/lucascullen/GitHub/horse-link/contracts.horse.link/node_modules/@ethersproject/abi/src.ts/coders/array.ts:71:19
    at Array.forEach (<anonymous>)
    at pack (/home/lucascullen/GitHub/horse-link/contracts.horse.link/node_modules/@ethersproject/abi/src.ts/coders/array.ts:54:12)
    at TupleCoder.encode (/home/lucascullen/GitHub/horse-link/contracts.horse.link/node_modules/@ethersproject/abi/src.ts/coders/tuple.ts:54:20)
    at AbiCoder.encode (/home/lucascullen/GitHub/horse-link/contracts.horse.link/node_modules/@ethersproject/abi/src.ts/abi-coder.ts:111:15)
  Error: ERROR processing /home/lucascullen/GitHub/horse-link/contracts.horse.link/deploy/040_Vault.ts:
  Error: invalid BigNumber value (argument="value", value=undefined, code=INVALID_ARGUMENT, version=bignumber/5.7.0)
      at Logger.makeError (node_modules/@ethersproject/logger/src.ts/index.ts:269:28)
      at Logger.throwError (node_modules/@ethersproject/logger/src.ts/index.ts:281:20)
      at Logger.throwArgumentError (node_modules/@ethersproject/logger/src.ts/index.ts:285:21)
      at Function.BigNumber.from (node_modules/@ethersproject/bignumber/src.ts/bignumber.ts:289:23)
      at NumberCoder.encode (node_modules/@ethersproject/abi/src.ts/coders/number.ts:25:27)
      at /home/lucascullen/GitHub/horse-link/contracts.horse.link/node_modules/@ethersproject/abi/src.ts/coders/array.ts:71:19
      at Array.forEach (<anonymous>)
      at pack (node_modules/@ethersproject/abi/src.ts/coders/array.ts:54:12)
      at TupleCoder.encode (node_modules/@ethersproject/abi/src.ts/coders/tuple.ts:54:20)
      at AbiCoder.encode (node_modules/@ethersproject/abi/src.ts/abi-coder.ts:111:15)
      at DeploymentsManager.executeDeployScripts (node_modules/hardhat-deploy/src/DeploymentsManager.ts:1222:19)
      at processTicksAndRejections (node:internal/process/task_queues:96:5)
      at runNextTicks (node:internal/process/task_queues:65:3)
      at listOnTimeout (node:internal/timers:528:9)
      at processTimers (node:internal/timers:502:7)
      at async DeploymentsManager.runDeploy (node_modules/hardhat-deploy/src/DeploymentsManager.ts:1052:5)
      at async Object.fixture (node_modules/hardhat-deploy/src/DeploymentsManager.ts:315:9)
      at async Context.<anonymous> (tests/registry.test.ts:26:19)

  3) Vault
       "before each" hook for "Should return the correct symbol":
     ERROR processing /home/lucascullen/GitHub/horse-link/contracts.horse.link/deploy/040_Vault.ts:
Error: invalid BigNumber value (argument="value", value=undefined, code=INVALID_ARGUMENT, version=bignumber/5.7.0)
    at Logger.makeError (/home/lucascullen/GitHub/horse-link/contracts.horse.link/node_modules/@ethersproject/logger/src.ts/index.ts:269:28)
    at Logger.throwError (/home/lucascullen/GitHub/horse-link/contracts.horse.link/node_modules/@ethersproject/logger/src.ts/index.ts:281:20)
    at Logger.throwArgumentError (/home/lucascullen/GitHub/horse-link/contracts.horse.link/node_modules/@ethersproject/logger/src.ts/index.ts:285:21)
    at Function.BigNumber.from (/home/lucascullen/GitHub/horse-link/contracts.horse.link/node_modules/@ethersproject/bignumber/src.ts/bignumber.ts:289:23)
    at NumberCoder.encode (/home/lucascullen/GitHub/horse-link/contracts.horse.link/node_modules/@ethersproject/abi/src.ts/coders/number.ts:25:27)
    at /home/lucascullen/GitHub/horse-link/contracts.horse.link/node_modules/@ethersproject/abi/src.ts/coders/array.ts:71:19
    at Array.forEach (<anonymous>)
    at pack (/home/lucascullen/GitHub/horse-link/contracts.horse.link/node_modules/@ethersproject/abi/src.ts/coders/array.ts:54:12)
    at TupleCoder.encode (/home/lucascullen/GitHub/horse-link/contracts.horse.link/node_modules/@ethersproject/abi/src.ts/coders/tuple.ts:54:20)
    at AbiCoder.encode (/home/lucascullen/GitHub/horse-link/contracts.horse.link/node_modules/@ethersproject/abi/src.ts/abi-coder.ts:111:15)
  Error: ERROR processing /home/lucascullen/GitHub/horse-link/contracts.horse.link/deploy/040_Vault.ts:
  Error: invalid BigNumber value (argument="value", value=undefined, code=INVALID_ARGUMENT, version=bignumber/5.7.0)
      at Logger.makeError (node_modules/@ethersproject/logger/src.ts/index.ts:269:28)
      at Logger.throwError (node_modules/@ethersproject/logger/src.ts/index.ts:281:20)
      at Logger.throwArgumentError (node_modules/@ethersproject/logger/src.ts/index.ts:285:21)
      at Function.BigNumber.from (node_modules/@ethersproject/bignumber/src.ts/bignumber.ts:289:23)
      at NumberCoder.encode (node_modules/@ethersproject/abi/src.ts/coders/number.ts:25:27)
      at /home/lucascullen/GitHub/horse-link/contracts.horse.link/node_modules/@ethersproject/abi/src.ts/coders/array.ts:71:19
      at Array.forEach (<anonymous>)
      at pack (node_modules/@ethersproject/abi/src.ts/coders/array.ts:54:12)
      at TupleCoder.encode (node_modules/@ethersproject/abi/src.ts/coders/tuple.ts:54:20)
      at AbiCoder.encode (node_modules/@ethersproject/abi/src.ts/abi-coder.ts:111:15)
      at DeploymentsManager.executeDeployScripts (node_modules/hardhat-deploy/src/DeploymentsManager.ts:1222:19)
      at processTicksAndRejections (node:internal/process/task_queues:96:5)
      at runNextTicks (node:internal/process/task_queues:65:3)
      at listOnTimeout (node:internal/timers:528:9)
      at processTimers (node:internal/timers:502:7)
      at async DeploymentsManager.runDeploy (node_modules/hardhat-deploy/src/DeploymentsManager.ts:1052:5)
      at async Object.fixture (node_modules/hardhat-deploy/src/DeploymentsManager.ts:315:9)
      at async Context.<anonymous> (tests/vault.test.ts:22:19)
```

File: contracts/VaultTimeLock.sol (hunk 1 above)
Comment (OptimusOpus) [reply contracts/VaultTimeLock.sol:35]: Nah

File: contracts/VaultTimeLock.sol (hunk 2 above)
Comment (OptimusOpus) [reply contracts/VaultTimeLock.sol:49]: This is implicitly done on the public function `withdraw` `require(assets <= maxWithdraw(owner), "ERC4626: withdraw more than max");`

**Review by brettvniekerk — DISMISSED** [review]

File: package.json [hunk 4]
```diff
@@ -4,10 +4,10 @@
   "private": true,
   "scripts": {
     "prepare": "husky install",
-    "hh:compile": "npx hardhat compile && npx hardhat deploy --network hardhat --tags underlying,oracle,registryToken,vault,market",
+    "hh:compile": "npx hardhat compile && npx hardhat deploy --network hardhat --tags underlying,oracle,registryToken,vault,vaultTimeLock,market",
```
Comment (bazmatic) [inline package.json:7]: `vaultTimeLock` tag no longer used so we can remove it from here and the other deployment commands

File: tests/vault.test.ts [hunk 5]
```diff
@@ -18,7 +19,12 @@ describe("Vault", () => {
 
 	beforeEach(async () => {
 		// Import deployments tagged with these values
-		const fixture = await deployments.fixture(["vault", "token", "market"]);
+		const fixture = await deployments.fixture([
+			"vault",
+			"vaultTimeLock",
```
[the remark this answers is not in the document]
Comment (OptimusOpus) [reply tests/vault.test.ts:24]: We are still using vault in this file

**Review by brettvniekerk — APPROVED** [review]

**Pull request merged**